RC smoothing retraining update - adds full support for CRSF

etracer65 opened the pull request:
Adds in flight monitoring of the rx frame rate and adapts the filters if the frame rate changes. Primarily to add support for Crossfire with its ability to switch from 150hz to 50hz (and back) under some circumstances. Will work with any protocol - not CRSF specific.   So if future receivers add the ability to switch frame rates dynamically the logic should support them.

If the current rx frame rate is more than +-20% from the previously detected rate, then the process will retrain for the next 50 samples as long as the rate continues to be outside the 20% tolerance. Once 50 samples are collected the new frame rate is updated and the filter cutoffs are adjusted. Only filters set with their cutoffs = 0 (auto) will be adjusted. There is a 2 second guard time after a successful update before retraining can start again to prevent rapid switching back and forth.

The logic is optimized to not perform any training if the filters are set to manual cutoffs. So there is an opportunity for advanced users to choose specific cutoffs and reduce the PID loop load slightly. However this is not recommended for Crossfire or other protocols that might change their rx frame rate.

Updated the output of the `rc_smoothing_info` cli command to match the revised logic.
```
# rc_smoothing_info
# RC Smoothing Type: FILTER
# Detected RX frame rate: 8.981ms
# Input filter type: BIQUAD
# Active input cutoff: 50hz (auto)
# Derivative filter type: BIQUAD
# Active derivative cutoff: 50hz (auto)
```

etracer65 commented:
@codecae @fujin Testing help for Crossfire would be appreciated.

The goal would be to make the rx switch from 150hz to 50hz while logging.  It's not necessary to be flying - bench testing is fine.  Please log with `debug_mode = RC_SMOOTHING_RATE`.  This will collect details on the frames received and the retraining cycle.

While I was testing I had a debug version that that added 5ms to the received rx frame rate based on a mode switch to simulate the rx changing rates while in flight.  So I'm confident that the detection/retraining/filter update logic works.  However we don't know if there are any edge cases or other strange behaviors for CRSF.  I had thought someone said that there might be a period of no frames when the rx starts to switch, and then it'll output 50hz.  That should be handled properly by the logic as it looks for "reasonable" frame intervals to ignore garbage data.

codecae commented on src/main/fc/rc_controls.h:
```diff
@@ -108,6 +107,23 @@ typedef enum {
+typedef struct rcSmoothingFilter_s {
+    bool filterInitialized;
+    biquadFilter_t filterBiquad[4];
+    pt1Filter_t filterPt1[4];
```
`filterInitialized` could possibly be replaced with a NULL check against the `filter_t` pointer, as well.

etracer65 replied:
Yes, you're correct and that may make sense to do at some point. But the plan was to have the ability to "plug in" different filtering types (like we investigated with moving-average) and those future types aren't likely to be compatible with filter_t.

Also I suspect that at some point we'll remove PT1 as BIQUAD does a so much better job.

etracer65 replied:
One additional type that shows promise is an "extrapolation" type of algoritm.

DieHertz replied:
`filter_t` is used as an opaque data type.
 Use union at least.

codecae replied:
IIRC, I believe `filter_t` is an opaque type that could be substituted for anything

codecae replied:
No matter :). Just saw another possible opportunity for flash savings.

DieHertz commented on src/main/fc/fc_rc.c:
```diff
@@ -277,25 +277,113 @@ int calcRcSmoothingCutoff(float avgRxFrameTime, bool pt1)
+    if ((rxConfig()->rc_smoothing_derivative_cutoff == 0) && (rxConfig()->rc_smoothing_derivative_type != RC_SMOOTHING_DERIVATIVE_OFF)) {
+        rcSmoothingData->derivativeCutoffFrequency = calcRcSmoothingCutoff(rcSmoothingData->averageFrameTimeUs, (rxConfig()->rc_smoothing_derivative_type == RC_SMOOTHING_DERIVATIVE_PT1));
+    }
+    if ((rcSmoothingData->derivativeCutoffFrequency != oldCutoff) || !rcSmoothingData->filterInitialized) {
```
This nested if can be replaced by:
```cpp
if (!rcSmoothingData->filterInitialized) {
  // initialize
} else if (rcSmoothingData->derivativeCutoffFrequency != oldCutoff) {
  // update
}
```
Same for non-derivative case.

etracer65 replied:
For the derivative case it makes sense.  I'm not sure about the input filter case.  To change that around I'd need to have a `switch` statement in each part of the `if !initialized` vs. an `if` in each `case`.  Either that or shift it up one level and have an entire !initialized vs. update `if` and replicate the `for` loop inside of each.

DieHertz replied:
Ok, I assumed they were similar

etracer65 commented:
Pushed another commit:

Saves about 200 bytes on F3.

Changed main storage structure to use union for the filters.

Renamed storage variable passed to the sub functions to avoid confusion with the global static.

Added whitespace to separate logical blocks and added additional comments to make the code more readable.

Restructured derivative filter initialization/update logic.

etracer65 commented:
Hmmm... minor test problem to fix.
```
Trivially initialized FAST_RAM variables found, use FAST_RAM_ZERO_INIT instead to save FLASH:
src/main/fc/fc_rc.c:80:static FAST_RAM rcSmoothingFilter_t rcSmoothingData;
make: *** [check-fastram-usage-correctness] Error 1
```

Question: Why doesn't that test run (or warn) when running the tests locally?

wind0r commented:
because that message is from make check-fastram-usage-correctness and not from a unit test.
it is an extra build step similar to make check-target-independence

DieHertz commented:
And it just helped you save some more FLASH :-D
